### ISSUE TITLE
Clusters get worse after running DEC

### ISSUE THREAD

**Issue by luckytoilet** [vlukiyanov/pt-dec#21]
I'm experiencing a strange issue, where running DEC makes the clustering worse. At first I encountered it in my project with real data, but to debug the issue, I managed to reproduce it with synthetic data as well.

In the attached notebook ([SyntheticDEC.zip](https://github.com/vlukiyanov/pt-dec/files/3388399/SyntheticDEC.zip)), I generate 1000 points in R^2 using a multivariate normal distribution, in two clusters. I then train an autoencoder to encode to 2 dimensions, then decode back the input. Finally, I run DEC for 100 epochs. The problem can be reproduced by running the notebook which has no external dependencies.

Initial hidden state:

![image](https://user-images.githubusercontent.com/123435/61166492-47831e80-a4fc-11e9-9516-810c8a70b1d1.png)

After 100 epochs of DEC:

![image](https://user-images.githubusercontent.com/123435/61166494-4eaa2c80-a4fc-11e9-8850-5c17f5866414.png)

After 200 epochs:

![image](https://user-images.githubusercontent.com/123435/61166582-fffd9200-a4fd-11e9-997d-5bda1f5cd085.png)

**Comment by vlukiyanov** [comment]
I think the accuracy on that dataset from just K-means should be near enough 100%, so its unlikely the DEC would improve it much; having run your example the accuracy remains constant at around 100% even as the representation becomes more 1-dimensional, so I guess the algorithm just converges to a lower dimensional representation.

If you make your data more overlapping by changing the `cov`, so for example weave through these changes [1] 

```python
data = np.concatenate([
  np.random.multivariate_normal(mean=(-0.5, 0), cov=((0.1, 0), (0, 0.1)), size=500),
  np.random.multivariate_normal(mean=(0.5, 0), cov=((0.1, 0), (0, 0.1)), size=500)
])
data = pd.DataFrame(data)
data.columns = ['x', 'y']
data['cluster'] = np.array([0]*500 + [1]*500)
```

And in each of your diagrams use `hue='cluster'` then things get quite interesting, initially the K-means at epoch 0 looks like:

![output0](https://user-images.githubusercontent.com/17338371/61171967-8e155f00-a576-11e9-922f-8c53beac7344.png)

and then changes to 

![output100](https://user-images.githubusercontent.com/17338371/61171971-953c6d00-a576-11e9-8348-3483fcc20576.png)

with a batch size of 32 the accuracy remains pretty constant. If you up it to 256 you see a drop in accuracy from the initial K-means, and then a slow rise back up. Possibly by changing the `cov` to be even more overlapping and introducing more clusters/dimensions things get even more interesting, I didn't try.

In short my understanding is the algorithm increases cluster accuracy, it doesn't necessarily improve the representation on a plot.

[1] I also had to change things like your function below

``python
def to_hidden(row):
    hidden = model.encoder(torch.Tensor(row[['x', 'y']]).cuda())
    #hidden = model(torch.Tensor(row).cuda())
    return pd.Series({'x': float(hidden[0]), 'y': float(hidden[1]), 'cluster': row.cluster})
```
and as (I guess) the `.iloc` and various Python-y things were slow I also decorated your `__getitem__` of `SyntheticDataset` with an in-memory `lru_cache` decorator. Those were the only changes I made, I think.

**Comment by luckytoilet** [comment]
I don't think that's quite right. Indeed, for the case with 2 clusters, the accuracy remains high, but with more than 2 clusters, the representation degenerates. For example, if we generate data with the following:

```
data = np.concatenate([
  np.random.multivariate_normal(mean=(-0.5, -0.5), cov=((0.04, 0), (0, 0.04)), size=500),
  np.random.multivariate_normal(mean=(0.5, -0.5), cov=((0.04, 0), (0, 0.04)), size=500),
  np.random.multivariate_normal(mean=(-0.5, 0.5), cov=((0.04, 0), (0, 0.04)), size=500),
  np.random.multivariate_normal(mean=(0.5, 0.5), cov=((0.04, 0), (0, 0.04)), size=500),
])
data = pd.DataFrame(data)
data.columns = ['x', 'y']
data['cluster'] = np.array([0]*500 + [1]*500 + [2]*500 + [3]*500)
```

Initial cluster accuracy is 0.9865:

![image](https://user-images.githubusercontent.com/123435/61172839-8a72df80-a558-11e9-9d71-b96194f1e61f.png)

After 100 epochs, cluster accuracy drops to around 0.5:

![image](https://user-images.githubusercontent.com/123435/61172845-a1b1cd00-a558-11e9-90ce-27cc756af770.png)

Furthermore nothing of this sort is observed in the original paper by [Xie et al. (2016)](https://arxiv.org/abs/1511.06335), figure 5:

![image](https://user-images.githubusercontent.com/123435/61172863-ec334980-a558-11e9-9575-4614a871deeb.png)

**Comment by vlukiyanov** [comment]
I think that screenshot refers to the MNIST dataset and is probably t-SNE or PCA plots, the autoencoders there will have higher hidden dimensions than your example. If I recall correctly the author actually suggests to use a fixed setup for the autoencoder where the encoder would have hidden layers `[input_dimension, 500, 500, 2000, 10]`, so that may be a thing to try (I don't remember if the 10 should be varied or not).

Another idea you could try to see what is reproduced when you prevent the autoencoder from overfitting, maybe add some noise to the encoder (e.g. add `nn.Dropout(0.2)` to your encoder) and greatly increase the various hidden dimensions.

If you want to compare to the results in the paper then one would have to reproduce the exact setup of the autoencoder and all other hyperparameter. I've tried to do that for the MNIST dataset.

For curiosity I took your example above and made the autoencoder [2,25,10] with a dropout. Then I used PCA plots since the hidden state is 10-dimensional. The data was a bit more overlapping, so

```python
data = np.concatenate([
  np.random.multivariate_normal(mean=(-0.5, -0.5), cov=((0.1, 0), (0, 0.1)), size=500),
  np.random.multivariate_normal(mean=(0.5, -0.5), cov=((0.1, 0), (0, 0.1)), size=500),
  np.random.multivariate_normal(mean=(-0.5, 0.5), cov=((0.1, 0), (0, 0.1)), size=500),
  np.random.multivariate_normal(mean=(0.5, 0.5), cov=((0.1, 0), (0, 0.1)), size=500),
])
```

And when I plotted the PCA plots over the epochs they went from

![output0](https://user-images.githubusercontent.com/17338371/61173614-9af17d00-a58d-11e9-9899-29c35f879b78.png)

to

![output100](https://user-images.githubusercontent.com/17338371/61173617-a04ec780-a58d-11e9-920e-0e49f2bb91f4.png)

after 100 epochs. The accuracy went up slightly from 85% to 88%.

Given the high number of hyperparameters and the instability of the model to these, I'm not sure what more there is to do unless you have identified a particular error in the code, or can check in another example which matches the paper setup exactly that we could then use for debugging. I believe there were a couple of others that could be done, I've only used MNIST. If you can do it for others maybe check them into the `examples` directory; without that it will be hard to track down if the implementation closely matches the paper or not.

--

Aside comment, I think in a few places in your notebook you should be detaching the Tensors that you get from the outputs (e.g. `.detach().cpu().numpy()`), actually you can actually do all the values in one go without a loop e.g. `model.encoder(torch.Tensor(data[['x', 'y']].values).cuda()).detach().cpu().numpy()`. Also I think your `epoch_callback` should switch the model to `eval` and then back to `train`.

**Comment by vlukiyanov** [comment]
Actually I was a bit curious, so I switched your model to the one suggested, so the code was like 

```python

class TrivialAE(nn.Module):
    def __init__(self):
        super(TrivialAE, self).__init__()

        # Input: B, 2
        self.encoder = nn.Sequential(
          nn.Linear(2, 500),
          nn.Tanh(),
          nn.Linear(500, 500),
          nn.Tanh(),
          nn.Linear(500, 2000),
          nn.Tanh(),
          nn.Linear(2000, 10),
          nn.Dropout(0.2)
        )

        # Input: B, 2
        self.decoder = nn.Sequential(
          nn.Linear(10, 2000),
          nn.Tanh(),
          nn.Linear(2000, 500),
          nn.Tanh(),
          nn.Linear(500, 500),
          nn.Tanh(),
          nn.Linear(500, 2),
        )

    def forward(self, x):
        x = self.encoder(x)
        x = self.decoder(x)
        return x
```

This was trained for 500 epochs. I also switched the DEC model to use a higher batch size of `256`. And then the PCA plots do look quite separated (I used the more overlapping inputs as in my previous comment).

They go from

![output_minus](https://user-images.githubusercontent.com/17338371/61173859-d5a8e480-a590-11e9-97ea-ed487675356d.png)

to

![output0](https://user-images.githubusercontent.com/17338371/61173863-df324c80-a590-11e9-8566-5957a1217536.png)

after a single epoch to

![output100](https://user-images.githubusercontent.com/17338371/61173865-e6f1f100-a590-11e9-8c3a-6c5d2a1d413a.png)

after 100. So perhaps the key is the method relies on a very particular autoencoder (perhaps it will work even better if the autoencoder matches the one in the paper fully).

**Comment by luckytoilet** [comment]
Interesting that it works in 10 dimensions and not in 2. However adding more dimensions and layers to the model doesn't seem like the way to go about debugging this, and only adds confounding variables and masks the underlying issue.

My understanding is that optimizing the KL divergence to the auxiliary distribution is supposed to pull the points closer to the cluster centroids. Also the centroids don't change in this implementation. So I would expect to see the points converge to tighter and tighter clusters around the initial 4 centroids as training progresses, even in 2 dimensions. This is not what we observe.

**Comment by vlukiyanov** [comment]
Out of curiosity I've tried with 2 and 4 in the setup of `[input_dimension, 500, 500, 2000, 2]` and `[input_dimension, 500, 500, 2000, 4]` and the results look comparable. In the MXNet implementation it seems like 10 is the default, see https://github.com/apache/incubator-mxnet/blob/master/example/deep-embedded-clustering/dec.py.

Your example is a lot shallower, so its hard to say why it might work differently to the suggested deeper network (t-SNE often diverges quite amusingly on examples that seem trivial, so it doesn't necessarily mean something is wrong). This sounds like it could be an open-ended research question rather than a coding one :)

It probably makes sense for you to try your example with the MXNet or Caffe implementation (see links in the README.md), since those are the official ones. If you notice a difference or can identify a bug please reopen the issue.
